### ISSUE TITLE
About the hyperparameters: "--losses", type=str, required=True, help="List of list of (loss_name, loss_params, bounds_name, bounds_params, fn, weight)"

### ISSUE THREAD

**Issue by wangyxxjtu** [LIVIAETS/semi_curriculum#1]
Hello, very nice project !
 I'm confused about the "--loss" hyperparameters, could you give some examples for such hyperparameter and what is it set in your experiment? could you provide examples for all "required " hyperparameters? 
Another problem is that the gt images produced by your script are 0 or 3, causing the assert error:assert sset(seg, list(range(C))), when the class number is set as 2. Why  it's not 0 or 1? 

Looking forward your reply, thank you!

**Comment by HKervadec** [comment]
Hey there,
Sorry for the delayed reply, and thanks for the interest in the project!

The `--loss` param is indeed quite difficult to use directly from the command line, I personally use `make` to automate the launches. See the file [learn_size.make](https://github.com/LIVIAETS/semi_curriculum/blob/master/learn_size.make). This one, I agree, is quite dense, as I used meta recipes to automate the experiments for a varying `n`.  The [make manual](https://www.gnu.org/software/make/manual/make.html) might be a good ressource if you want to dig deeper into that. 

I find makefiles great to deal with the dependencies between different parts of the training (in this case, getting the right subset of training images, training the auxiliary network, and then training the segmentation network with the trained auxiliary network). 


A quick example, pasted from the current head of my private repository (small differences might be present). I commented the output to explain the different parts: 
```shell
# make -n print the command that will be run, without actually running it
>>> make -f learn_size.make results/learn_size/semi_pred_size_40 -n 
# extract the acdc datset, and keep what is of interest for us. 
# Notice I have a slight bug that I need to fix, as it does that part twice.
rm -rf data/ACDC-2D-LS_tmp data/ACDC-2D-LS
PYTHONPATH=":." python3.7  preprocess/slice_acdc.py --source_dir="data/acdc/training" --dest_dir=data/ACDC-2D-LS_tmp --seed=1 --retain 25 --fold 0
python3.7 remap_values.py data/ACDC-2D-LS_tmp/train/gt "{3: 3, 2: 0, 1: 0, 0: 0}"
python3.7 remap_values.py data/ACDC-2D-LS_tmp/val/gt "{3: 3, 2: 0, 1: 0, 0: 0}"
mv data/ACDC-2D-LS_tmp data/ACDC-2D-LS
rm -rf data/ACDC-2D-LS_tmp data/ACDC-2D-LS
PYTHONPATH=":." python3.7  preprocess/slice_acdc.py --source_dir="data/acdc/training" --dest_dir=data/ACDC-2D-LS_tmp --seed=1 --retain 25 --fold 0
python3.7 remap_values.py data/ACDC-2D-LS_tmp/train/gt "{3: 3, 2: 0, 1: 0, 0: 0}"
python3.7 remap_values.py data/ACDC-2D-LS_tmp/val/gt "{3: 3, 2: 0, 1: 0, 0: 0}"
mv data/ACDC-2D-LS_tmp data/ACDC-2D-LS

# Create the sub validation set, with 5 patients
cp -r data/ACDC-2D-LS/val data/ACDC-2D-LS/val_r_20_tmp
for f in `ls data/ACDC-2D-LS/val_r_20_tmp` ; do \
	echo $f ; \
	for im in `ls data/ACDC-2D-LS/val_r_20_tmp/img | cut -d '_' -f 1 | sort | uniq | head -n 20` ; do \
		rm data/ACDC-2D-LS/val_r_20_tmp/$f/$im* ; \
	done \
done
mv data/ACDC-2D-LS/val_r_20_tmp data/ACDC-2D-LS/val_r_20

# Create the sub training set, with 40 labeled patients 
cp -r data/ACDC-2D-LS/train data/ACDC-2D-LS/train_b_40_tmp
for f in `ls data/ACDC-2D-LS/train_b_40_tmp` ; do \
	echo $f ; \
	for im in `ls data/ACDC-2D-LS/train_b_40_tmp/img | cut -d '_' -f 1 | sort | uniq | tail -n +41` ; do \
		rm data/ACDC-2D-LS/train_b_40_tmp/$f/$im* ; \
	done \
done
mv data/ACDC-2D-LS/train_b_40_tmp data/ACDC-2D-LS/train_b_40

# Train the auxiliary network with 40 labeled patients
rm -rf results/learn_size/pred_size_40_tmp
mkdir -p results/learn_size/pred_size_40_tmp
printenv > results/learn_size/pred_size_40_tmp/env.txt
git diff > results/learn_size/pred_size_40_tmp/repo.diff
git rev-parse --short HEAD > results/learn_size/pred_size_40_tmp/commit_hash
python3.7  train_regression.py --n_class 4 --base-width 4 --cardinality 32 --save_folder results/learn_size/pred_size_40_tmp --data_root data/ACDC-2D-LS --epc 500 \
	--train_subfolder train_b_40 --val_subfolder val_r_20 --equalize --idc 3 --batch_size 10 --in_memory
mv results/learn_size/pred_size_40_tmp results/learn_size/pred_size_40

# Create the sub validation set, with the remaining 20 patients
cp -r data/ACDC-2D-LS/val data/ACDC-2D-LS/val_20_tmp
for f in `ls data/ACDC-2D-LS/val_20_tmp` ; do \
	echo $f ; \
	for im in `ls data/ACDC-2D-LS/val_20_tmp/img | cut -d '_' -f 1 | sort | uniq | tail -n +21` ; do \
		rm data/ACDC-2D-LS/val_20_tmp/$f/$im* ; \
	done \
done
mv data/ACDC-2D-LS/val_20_tmp data/ACDC-2D-LS/val_20

# Create the sub trainig set, with the remaining (75 - 40) unlabeled patients)
cp -r data/ACDC-2D-LS/train data/ACDC-2D-LS/train_r_40_tmp
for f in `ls data/ACDC-2D-LS/train_r_40_tmp` ; do \
	echo $f ; \
	for im in `ls data/ACDC-2D-LS/train_r_40_tmp/img | cut -d '_' -f 1 | sort | uniq | head -n 40` ; do \
		rm data/ACDC-2D-LS/train_r_40_tmp/$f/$im* ; \
	done \
done
mv data/ACDC-2D-LS/train_r_40_tmp data/ACDC-2D-LS/train_r_40

# Actually train the segmentation network, with 40 labeled patients, and 35 unlabeled patients.
# It uses the trained auxiliary network from earlier
make -f learn_size.make train_template DEST=results/learn_size/semi_pred_size_40 DEPS="data/ACDC-2D-LS/val_20 data/ACDC-2D-LS/train_b_40 data/ACDC-2D-LS/train_r_40 results/learn_size/pred_size_40"
make[1]: Entering directory '/home/hoel/code/constrained_cnn'
rm -rf results/learn_size/semi_pred_size_40_tmp
mkdir -p results/learn_size/semi_pred_size_40_tmp
printenv > results/learn_size/semi_pred_size_40_tmp/env.txt
git diff > results/learn_size/semi_pred_size_40_tmp/repo.diff
git rev-parse --short HEAD > results/learn_size/semi_pred_size_40_tmp/commit_hash
python3.7  main.py --dataset=data/ACDC-2D-LS/ --batch_size=1 --group --schedule \
	--n_epoch=100 --workdir=results/learn_size/semi_pred_size_40_tmp --csv=metrics.csv --n_class=4 --metric_axis 3 \
	--grp_regex="(patient\d+_\d+_\d+)_\d+" --validation_folder val_20 --in_memory --network=ENet --losses="[[('CrossEntropy', {'idc': [0, 3]}, None, None, None, 1)], [('NaivePenalty', {'idc': [3]}, 'PredictionBounds', {'margin': 0.10, 'mode': 'percentage', 'net': 'results/learn_size/pred_size_40/model.pkl'}, 'soft_size', 1e-2)]]" --training_folders train_b_40 train_r_40 --val_loader_id 0 --folders="[[('img', png_transform, False), ('gt', gt_transform, True)]+[('gt', gt_transform, True)], [('img', png_transform, False), ('gt', gt_transform, True)]+[('gt', dummy_gt, False)]]"
mv results/learn_size/semi_pred_size_40_tmp results/learn_size/semi_pred_size_40
make[1]: Leaving directory '/home/hoel/code/constrained_cnn'
rm data/ACDC-2D-LS/train_b_40 data/ACDC-2D-LS/val_r_20
```

I usually start from an existing makefile, and then modify it depending on my needs (I use the same codebase for most of my papers).

---
>  Another problem is that the gt images produced by your script are 0 or 3, causing the assert error:assert sset(seg, list(range(C))), when the class number is set as 2. Why it's not 0 or 1?

This is a decision I took a while ago, since ACDC has actually 4 classes (background, right ventricle, myocardium, and left ventricle). While I remap RV and myocardium to background (`{1,2} -> 0`), I prefer to keep the label for LV at 3 ; which makes comparison and metrics computation against the original dataset easier. 

Therefore, I train my network for 4 classes, but bother only supervising classes `0` (background) and `3` (LV). This is what `'CrossEntropy', {'idc': [0, 3]}` does. The extra computation overhead (wider last layer) is negligible. 

Feel free to update that part, it is mostly a matter of personal taste.

Let me know if you need any help. While my code is quite easy to re-run verbatim, I am aware that there is a lot of information to digest if you want to modify it.

Best,

Hoel

**Comment by wangyxxjtu** [comment]
Thank you for your detailed reply. My difficulties are mainly from that I'm NOT familiar with the grammar of the makefile, I will continue to read the learn_size.make file after learning some necessary grammars. Anyway, Thank you!

In fact, I also focus on your another project "SizeLoss_WSS" (https://github.com/LIVIAETS/SizeLoss_WSS), it is also a great work. When I run the acdc.make, I found the main.py file would run multiple times with different configs like "fs", 'partial', 'upper', 'loose' , 'pathak_precision', etc. Unfortunately, I did not find some comments for these configs, leading to some confusion for me.  Could you explain the exact meaning for these configs, and how are the correspondings between performance reported in  the paper (Constrained-CNN Losses for Weakly Supervised Segmentation) and these configs ?

Thank you very much,
Looking forward your reply.

**Comment by HKervadec** [comment]
For Table 1 in the paper:

Table 1 name | Recipe in makefile
-----|-----
Partial CE | `partial`
CE + tags (lag prop) | `pathak_loose`
Partial CE + tags (direct loss) | `loose`
CE + tags + size (common) (lag prop) | `pathak_595`
Partial CE + tags + size (common) (direct loss) | `size_595`
CE + tags + size (individual) (lag prop) | `pathak_precise`
Partial CE + tags + size (individual) (direct loss) | `presize` 
Partial CE + 3d size (individual) (direct loss) | `3d_sizeloss_random` 
Full supervision | `fs`

You can usually figure it out by checking the options in the `--loss` parameters. `upper` usually means that I have only an upper bound, i.e. the lower bound is 0 (used in the ablation study in Table 4)

**Comment by wangyxxjtu** [comment]
Ok. Thank you very much for your reply!
